registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+		return "", err
+	}
+	if err := fn(page.Tags); err != nil {
+		return "", err
+	}
+
+	return parseLink(resp)
+}
+
+// UpEdges returns the manifest descriptors directly referencing the given
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) UpEdges(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	var res []ocispec.Descriptor
+	if err := r.Referrers(ctx, desc, func(referrers []artifactspec.Descriptor) error {
+		for _, referrer := range referrers {
+			res = append(res, descriptor.ArtifactToOCI(referrer))
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Referrers returns the manifest descriptors directly referencing the given
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, fn func(referrers []artifactspec.Descriptor) error) error {
+	// TODO(shizhMSFT): filter artifact type
+	url := fmt.Sprintf("%s/manifests/%s/referrers", r.artifactEndpoint(), desc.Digest)
+	if r.ReferrerListPageSize > 0 {
+		url = fmt.Sprintf("%s?n=%d", url, r.ReferrerListPageSize)
+	}
+
+	var err error
+	for err == nil {
+		url, err = r.referrers(ctx, desc, fn, url)
+	}
+	if err != errNoLink {
+		return err
+	}
+	return nil
+}
+
+// referrers returns a single page of the manifest descriptors directly
+// referencing the given manifest descriptor with the next link.
+func (r *Repository) referrers(ctx context.Context, desc ocispec.Descriptor, fn func(referrers []artifactspec.Descriptor) error, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := r.Client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", parseErrorResponse(resp)
+	}
+
+	var page struct {
+		References []artifactspec.Descriptor `json:"references"`
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+		return "", err
+	}
+	if err := fn(page.References); err != nil {
+		return "", err
+	}
+
+	return parseLink(resp)
+}
+
+// scheme returns HTTP scheme used to access the remote registry.
+func (r *Repository) scheme() string {
+	if r.PlainHTTP {
+		return "http"
+	}
+	return "https"
+}
+
+// endpoint returns the base endpoint of the remote registry.
+func (r *Repository) endpoint() string {
+	return fmt.Sprintf("%s://%s/v2/%s", r.scheme(), r.Reference.Host(), r.Reference.Repository)
+}
+
+// artifactEndpoint returns the base endpoint of the remote registry for ORAS
+// artifacts.
+func (r *Repository) artifactEndpoint() string {
```

Comment:
Should we consider another file or a set of constants for all the path building. For e.g. you have blobs/manifests/tags and not `artifacts` all hard coded here. If we change the path I think it would go maybe have something like 
`RepoArtifactReferrersGetURLFormat` or `{entity}{operation}UriFormat` for e.g.

Reply:
I think this is a tricky part. We don't have a `v1.0.0` [artifact-spec](https://github.com/oras-project/artifacts-spec) yet. The format itself can change in a totally different way. Once the spec is finalized, the path is fixed and there will not be any demand on changing the path.

Therefore, I would like not to expose something like `RepoArtifactReferrersGetURLFormat`. If we do want this, I think the `artifacts-spec` repo is a better home.

Reply:
My concern is more about the large number of hardcoded string paths. Even internal constants would be better in a single location don’t have to expose them.

Reply:
Currently, a URL is built like

```golang
url := fmt.Sprintf("%s/manifests/%s", r.endpoint(), reference)
```

If we move to format like

```golang
manifestPathFormat := "%s://%s/v2/%s/manifests/%s"
url := fmt.Sprintf(manifestPathFormat, r.scheme(), r.Reference.Host(), r.Reference.Repository, reference)
```
It would be worse. Since the definition of `manifestPathFormat` is at somewhere else, it is harder to check if the correct number of arguments has been passed in to match the format.

If we do something like

```golang
const manifestPathFormat := "%s://%s/v2/%s/manifests/%s"

func buildManifesetPath(scheme, host, repository, reference string) string {
    return fmt.Sprintf(manifestPathFormat, scheme, host, repository, reference)
}
```

then do 

```golang
url := buildManifesetPath(r.scheme(), r.Reference.Host(), r.Reference.Repository, reference)
```

It might be good but ... is it an overkill?

Here are the status of other products:
- `distribution`: it has URL builders in the `distribution/registry/api/v2` package.
- `containerd`: paths are hard coded. Parts joined using `strings.Join`.
- `crane`: paths are hard coded. Parts joined using `fmt.Sprintf`.

Reply:
Yes the URL Builder pattern is exactly what I was hoping we would follow. 
https://github.com/distribution/distribution/blob/main/registry/api/v2/urls.go

The reason is that I would recommend the .NET SDK also to follow the routes from here as a baseline so having a single URLBuilder enables us to modify the routes in one place with capabilities like extension check and ORAS paths etc.

Reply:
I will update the PR to adapt the pattern.

Reply:
Update with 51695f3 to adapt the build URL pattern.